WSL refactor to handle multiple distros and use JB WSL API

sh41 opened the pull request:
Fixes test suite failures and resolves WSL distribution execution issues.

## Part 1: Test Suite Fixes

Fixes multiple test failures caused by path handling issues on Windows and incorrect test assertions, unblocking work on WSL refactoring.

1. **`PathUtils.getRelativePath()` - Fix cross-platform path handling**
    - Normalize paths to system-independent format before calculating relative paths
    - Fixes failures on Windows where VFS uses forward slashes but `File.separatorChar` is backslash

2. **`MiseServiceTest` - Update for VFS structure and add environment testing**
    - Fixed path expectations to include `/src/` prefix from test VFS
    - Refactored to parameterized tests: 24 tests (8 tasks × 3 environments)
    - Validates environment-specific config loading (e.g., `mise.test.toml` only loads when `MISE_ENV=test`)
    - Tests both presence and absence of tasks based on environment

3. **`MiseTomlTableTaskTest` - Fix dependency assertion**
    - Updated to access nested dependency array: `[0]` → `[0][0]`

4. **`MiseTomlTaskRunConfigurationProducerTest` - Fix test setup**
    - Fixed TOML comment syntax: `//` → `#`
    - Updated `createConfigurationContext()` to wrap element in `PsiLocation`

## Part 2: Multi-WSL Support and Major Refactoring

This part addresses #395 by adding comprehensive WSL support for projects hosted in multiple WSL distributions, along with significant architectural improvements to command execution, caching, and configuration management.

### 🎯 Core Features

**Multi-WSL Distribution Support**
- Projects can now be opened from different WSL distributions simultaneously
- Automatic detection and routing of mise commands to the correct WSL context based on project location
- UNC path support (`\\wsl.localhost\Distribution\...`) with proper context-aware execution
- IntelliJ Eel framework integration for reliable WSL command execution

**Centralized Executable Management**
- New `MiseExecutableManager` provides single source of truth for mise executable resolution
- Clear priority order: project settings → application settings → auto-detection
- Per-project executable path configuration for container/WSL-specific setups
- Unified UI with "project-only" checkbox pattern (inspired by Git VCS settings)
- Auto-detected path shown as placeholder for better discoverability

**Intelligent Command Caching**
- Caches `mise env`, `mise ls`, and `mise version` to prevent excessive subprocess calls
- Smart invalidation: watches actual config files and invalidates only when they change
- Stampede protection: concurrent requests for same data execute only once
- Thread-safe with per-key locking
- Broadcast-based invalidation via message bus topics (simpler and more reliable)
- File rename detection (e.g., `mise.test.yml` → `mise.test.toml` triggers invalidation)

**Granular Environment Variable Injection**
- Per-feature toggles for mise environment variable usage:
  - Run configurations
  - Database authentication  
  - Nx commands
  - Global command line injection (opt-in catch-all for terminal, external tools, build tools)
- Marker-based double-injection prevention
- Mise executable exclusion to prevent infinite recursion

### 🐛 Major Fixes

**Cross-Platform Path Handling**
- `PathUtils.getRelativePath()` now normalizes to system-independent format before calculation
- Resolves Windows path separator issues (VFS uses `/` but `File.separatorChar` is `\`)

**Context-Aware Version Detection**
- `getMiseVersion()` now instance method with project context awareness
- Per-executable version caching prevents incorrect flag usage (`--env` vs `--profile`)
- Fixes issue where Windows mise version was used for WSL project operations

**Notification Improvements**
- Debounced untrusted config notifications (was showing 4+ duplicate notifications)
- Per-file notification tracking prevents spam
- Removed annoying "executable not found" notification when mise correctly available in PATH

**Test Suite Improvements**
- Fixed failing tests with parameterized `MiseServiceTest` (24 scenarios: 8 tasks × 3 environments)
- Virtual file system path handling (`/src/` prefix from test fixtures)
- Corrected TOML comment syntax (`//` → `#`)
- Fixed nested dependency array access in `MiseTomlTableTaskTest`
- WSL-specific tests only run on Windows platforms
- Converted to `LightPlatformTestCase` for proper IDE testing support

### 🔧 Build & Dependencies

- Migrated to new `intellijPlatform.create` API (fixes deprecation warnings)
- Explicit Jackson Kotlin module dependency (aligns with IntelliJ Platform 2025.1)
- Git4Idea plugin dependency declaration (resolves `vcs.envCustomizer` warnings)
- Platform-specific run configurations for easier testing (PyCharm Community, multiple IDEA Ultimate versions)
- Replaced obsolete `runPluginVerifier` references with `verifyPlugin`

### 📊 Statistics

- **52 files changed**: 1,374 insertions(+), 745 deletions(-)
- **33 commits** covering test fixes, WSL support, caching, and configuration refactoring
- **Major code reduction**: 300+ lines removed through architectural simplification
- **New utilities**: `StampedeProtectedCache`, `guessMiseProjectPath`, `MiseExecutableManager`

### ✅ Testing Status

- ✅ All existing tests passing
- ✅ macOS compatibility verified (@134130)
- ✅ Windows → WSL multi-distribution testing in progress (@sh41)
- ⏳ Awaiting CI pipeline testing before merge

Built package for early adopters: [mise-5.13.0-multi-wsl-20260110.zip](https://github.com/user-attachments/files/24538685/mise-5.13.0-multi-wsl-20260110.zip)

sh41 commented:
@gemini-code-assist review

sh41 commented:
/gemini review

134130 commented on modules/core/src/main/kotlin/com/github/l34130/mise/core/setting/MiseConfigurable.kt:
```diff
@@ -47,23 +56,33 @@ class MiseConfigurable(
-                            Not installed? Visit the <a href='https://mise.jdx.dev/installing-mise.html'>mise installation</a></br>
-                            For WSL: <code>wsl.exe -d DistroName mise</code> or <code>\\wsl.localhost\DistroName\usr\bin\mise</code></br>
+                            Optional: Override mise executable path for all projects.</br>
+                            Leave empty to auto-detect 'mise' from PATH or common locations (recommended).</br>
```
I attempted to explicitly specify the executable file path.

This is because other IDEA settings automatically detect tools and add them to the dropdown list.

For a unified user experience, automatically detecting tools and adding them to the text field seems preferable. Therefore, I added a notification at `MiseProjectSettings.kt:L40`. (If it doesn't work, it's a bug.)

sh41 replied:
For some reason I only noticed your new commits after I posted the above comment.  I see you've added a checkbox-based project override. I like this approach, similar to the `git` model  - it makes the override explicit and opt-in. 

However, I noticed the fallback chain may need adjustment - currently when the checkbox is unchecked, it appears to skip the application setting entirely and go straight to PATH. Is this intentional?

I think that application level setting makes sense for the non-WSL scenarios, i.e. the user is on macOS or Linux or Windows without WSL (although that would be hard to detect and could change). When WSL is involved I think that application level has potential to break things. Maybe there should be two different configuration UIs, one for macOS/Linux and one for Windows? 


Let me know your thoughts 👍

sh41 commented:
I've been using this version for a while today and noticed that there are a lot of duplicated calls to `env`, `version` & `ls`. So I've added caching for those commands (mise env, mise ls, mise task ls) which stops us from hammering the CLI constantly.

What it does:
Stops duplicate calls - if multiple things ask for the same data at once, only one actually runs the command
Auto-invalidates smartly - watches the actual config files being used and only re-runs when they change
Thread-safe - uses per-key locks so concurrent requests don't step on each other
Result: Way fewer mise subprocess calls, especially on start up when everything tries to load at once.

Here's an extract from the log showing the cache hits, then I edit the `mise.toml` which invalidates the cache, so it refreshes and then carries on responding from cache: 
```
2025-12-30 15:23:05,230 [   7501]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cached: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1: (invalidation=ON_CONFIG_CHANGE)
2025-12-30 15:23:05,230 [   7501]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit after lock: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:05,230 [   7501]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit after lock: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:05,230 [   7501]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit after lock: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:05,231 [   7502]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit after lock: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:05,241 [   7512]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:05,241 [   7512]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:05,242 [   7513]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cached: ls://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1: (invalidation=ON_CONFIG_CHANGE)
2025-12-30 15:23:05,242 [   7513]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit after lock: ls://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:05,242 [   7513]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit after lock: ls://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:05,251 [   7522]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:05,291 [   7562]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:05,293 [   7564]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:05,299 [   7570]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:05,310 [   7581]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:05,324 [   7595]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:05,464 [   7735]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:05,490 [   7761]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:05,592 [   7863]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:05,598 [   7869]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:05,603 [   7874]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:05,793 [   8064]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:06,295 [   8566]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:06,303 [   8574]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:23:08,335 [  10606]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:24:53,136 [ 115407]   INFO - #com.github.l34130.mise.core.command.MiseCommandCache - File changed: //wsl.localhost/Ubuntu-22.04/home/user/workspace/project1/mise.toml, invalidating 2 cache entries
2025-12-30 15:24:53,137 [ 115408]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Invalidated: ls://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:24:53,137 [ 115408]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Invalidated: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:24:53,526 [ 115797]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache miss or invalid: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:24:53,526 [ 115797]   FINE - #com.github.l34130.mise.core.command.MiseCommandLine - ==> [COMMAND] Starting command execution (workDir: //wsl.localhost/Ubuntu-22.04/home/user/workspace/project1, params: [env, --json])
2025-12-30 15:24:53,526 [ 115797]   FINE - #com.github.l34130.mise.core.command.MiseCommandLine - ==> [EXECUTABLE] Using 'mise' from PATH (context: WSL, workDir: //wsl.localhost/Ubuntu-22.04/home/user/workspace/project1)
2025-12-30 15:24:53,526 [ 115797]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: version:mise
2025-12-30 15:24:53,526 [ 115797]   FINE - #com.github.l34130.mise.core.command.MiseCommandLine - ==> [EXECUTABLE] Using 'mise' from PATH (context: WSL, workDir: //wsl.localhost/Ubuntu-22.04/home/user/workspace/project1)
2025-12-30 15:24:53,526 [ 115797]   FINE - #com.github.l34130.mise.core.command.MiseCommandLine - ==> [PATH TRANSFORM] isWslContext: true (exec: mise, workDir: //wsl.localhost/Ubuntu-22.04/home/user/workspace/project1)
2025-12-30 15:24:53,526 [ 115797]   FINE - #com.github.l34130.mise.core.command.MiseCommandLine - ==> [EXEC] mise env --json (workDir: \\wsl.localhost\Ubuntu-22.04\home\user\workspace\project1)
2025-12-30 15:24:54,412 [ 116683]   FINE - #com.github.l34130.mise.core.command.MiseCommandLine - Command executed successfully. (command=mise [env, --json])
2025-12-30 15:24:54,412 [ 116683]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cached: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1: (invalidation=ON_CONFIG_CHANGE)
2025-12-30 15:24:54,428 [ 116699]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:24:54,438 [ 116709]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:28:22,228 [ 324499]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:28:22,234 [ 324505]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:28:22,334 [ 324605]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:28:22,351 [ 324622]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:28:22,365 [ 324636]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
2025-12-30 15:28:22,383 [ 324654]   FINE - #com.github.l34130.mise.core.command.MiseCommandCache - Cache hit: env://wsl.localhost/Ubuntu-22.04/home/user/workspace/project1:
```

sh41 commented:
Hi @134130, I've done some more refactoring. This is the Mise Executable area on Windows: 

<img width="802" height="203" alt="image" src="https://github.com/user-attachments/assets/d8a2653b-7f60-47ec-9af1-ce2a6b30c2be" />

And this is on Windows with a WSL project: 
<img width="699" height="228" alt="image" src="https://github.com/user-attachments/assets/e388ddae-81a0-4119-b404-242d33e40b0e" />

And this is on Linux: 
<img width="803" height="214" alt="image" src="https://github.com/user-attachments/assets/e8fabe14-6b6e-4593-bf3c-9789d1f02f25" />

The project home is essential to determining the correct WSL Distro to run in, so I added a helper `guessMiseProjectPath` which handles that and use it everywhere that the command line logic is called. 

I've also refactored out the caching logic to make it reusable and implemented caching of the executable path so that it is absolutely consistent in across every usage location and when displayed in the UI. 

Given the size of this change, I think this needs some real world testing before it is considered done. If you have a canary channel that would be a good place to put it, or there's a built version here that people can install and test with. 

[mise-5.13.0-multi-wsl.zip](https://github.com/user-attachments/files/24388386/mise-5.13.0-multi-wsl.zip)

I don't have macOs available to test this on, but will be at work and using it on a Windows -> multiple WSL setup from Monday next week. 

Thanks again for the great plugin & apologies for the huge PR!

@gemini-code-assist review

sh41 commented:
I made it so that the tests that rely on WSL only run on Windows platforms. 

I also changed the way that the executable is queried in WSL. 

New build here: [mise-5.13.0-multi-wsl.zip](https://github.com/user-attachments/files/24396758/mise-5.13.0-multi-wsl.zip)

@gemini-code-assist review

134130 commented:
macOS looks fine

sh41 commented on modules/core/src/main/kotlin/com/github/l34130/mise/core/command/MiseExecutableManager.kt:
```diff
@@ -0,0 +1,488 @@
+            }
+        } catch (e: Exception) {
+            logger.debug("Failed 'where' command", e)
+        }
```
Removed this entirely.

134130 reviewed (approved):
Incredible. I really appreciate to you!
